frontend_server/lib/server.dart

```diff
@@ -0,0 +1,160 @@
+import 'package:kernel/target/flutter_fasta.dart';
+import 'package:kernel/target/targets.dart';
+import 'package:usage/uuid/uuid.dart';
+
+ArgParser argParser = new ArgParser(allowTrailingOptions: true)
+  ..addFlag('train',
+      help: 'Run through sample command line to produce snapshot',
+      negatable: false)
+  ..addFlag('incremental',
+      help: 'Run compiler in incremental mode', defaultsTo: false)
+  ..addOption('sdk-root',
+      help: 'Path to sdk root',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk')
+  ..addOption('platform-kernel-dill',
+      help: 'Path to platform kernel dill file',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk/platform.dill');
+
+String usage = '''
+Usage: server [options] [input.dart]
+
+If input dart source code is provided on the command line, then the server 
+compiles it, generates dill file and exits.
+If no input dart source is provided on the command line, server waits for 
+instructions from stdin.
+
+Instructions:
+- compile <input.dart>
+- recompile <boundary-key>
+<invalidate-file1.dart>
+<invalidate-file2.dart>
+...
+<boundary-key>
+- accept
+- reject
+- quit
+
+Output:
+- result <boundary-key>
+<compiler output>
+<boundary-key> [<output.dill>]
+
+Options:
+${argParser.usage}
+''';
+
+enum State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
+
+starter(List<String> args) async {
+  ArgResults options = argParser.parse(args);
+  if (options['train']) {
+    return 0;
+  }
+
+  if (options.rest.isNotEmpty) {
+    await compile(options.rest[0], options);
+    return 0;
+  }
+
+  State state = State.READY_FOR_INSTRUCTION;
+  String boundaryKey;
+  print('Frontend server is ready.');
+  stdin
+      .transform(UTF8.decoder)
+      .transform(new LineSplitter())
+      .listen((String string) async {
+    switch (state) {
+      case State.READY_FOR_INSTRUCTION:
+        if (string.startsWith('compile ')) {
+          String filename = string.substring('compile '.length);
+          await compile(filename, options);
+        } else if (string.startsWith('recompile ')) {
+          boundaryKey = string.substring('recompile '.length);
+          state = State.RECOMPILE_LIST;
+        } else if (string == 'accept') {
+          ikg.acceptLastDelta();
+        } else if (string == 'reject') {
+          ikg.rejectLastDelta();
+        } else if (string == 'quit') {
+          exit(0);
+        }
+        break;
+      case State.RECOMPILE_LIST:
+        if (string == boundaryKey) {
+          recompileDelta();
+          state = State.READY_FOR_INSTRUCTION;
+        } else {
+          ikg.invalidate(Uri.parse(string));
+        }
+        break;
+    }
+  });
+}
+
+IncrementalKernelGenerator ikg;
+String filename;
+
+compile(String filename, ArgResults options) async {
+  var program;
+  String boundaryKey = new Uuid().generateV4();
+  print("result $boundaryKey");
+  if (options['incremental']) {
+    final CompilerOptions compilerOptions = new CompilerOptions()
```

Comment:
minor nit: since there is so much shared between this and the options below, consider moving this up above the if-else (especially that with my suggestions further below, the only different might be a single option):

```dart
var sdkRootPath = options['sdk-root'];
if (!sdkRootPath.endsWith('/')) sdkRootPath = '$sdkRootPath/';
var sdkRoot = Uri.base.resolve(sdkRootPath);
final CompilerOptions compilerOptions = new CompilerOptions()
    ..sdkRoot = sdkRoot
    ..strongMode = false
    ..target = new FlutterTarget(new TargetFlags())
    ..onError = (CompilationMessage message) {
￼        print(message);
￼      };
if (options['incremental']) {
  ikg = await ...;
} else {
  // TODO(aam): remove linked depenendencies once platform is directly embedded in the
  // VM snapshot and issue XYZ is fixed.
  compilerOptions.linkedDependencies = <Uri>[sdkRoot.resolve('platform.dill')];
  program = await ...;
}